accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+      intent. Idiomatic gestures, such as guarding the call with a platform
+      check or marking a method as platform-specific, automatically suppress
+      these diagnostics. However, explicit suppressions via the usual means
+      (`NoWarn`, `#pragma`) will also work.
+* The analyzer helps library authors to express their intent so that their
+  consumers can also benefit from these diagnostics.
+    - Library authors can annotate their platform-specific APIs without taking a
+      dependency.
+    - Library consumers don't have to install a NuGet package to get diagnostics
+      when using platform-specific APIs
+
+### Non-Goals
+
+* Modelling partially portable APIs (that is APIs are applicable to more than
+  one platform but no all platforms). Those should Be modeled as capability
+  APIs, as as `RuntimeInformation.IsDynamicCodeSupported` for ref emit. This
+  would be a separate analyzer.
+* Shipping the platform-specific analyzer and/or annotations out-of-band.
+
+## Design
+
+### Attribute
+
+OS bindings and platforms-specific BCL APIs will be annotated using a set of
+custom attributes:
+
+```C#
+namespace System.Runtime.InteropServices
+{
+    public partial struct OSPlatform
+    {
+        // Existing properties
+        // public static OSPlatform FreeBSD { get; }
+        // public static OSPlatform Linux { get; }
+        // public static OSPlatform OSX { get; }
+        // public static OSPlatform Windows { get; }
+        public static OSPlatform Android { get; }
+        public static OSPlatform iOS { get; }
+    }
+
+    public partial static class RuntimeInformation
+    {
+        // Existing API
+        // public static bool IsOSPlatform(OSPlatform osPlatform);
+
+        // Check for the OS with a >= version comparison
+        // Used to guard APIs that were added in the given OS release.
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build, int revision);
+
+        // Allows checking for the OS with a < version comparison
+        // Used to guard APIs that were obsoleted or removed in the given OS release.
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major);
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major, int minor);
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major, int minor, int build);
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major, int minor, int build, int revision);
+    }
+}
+
+namespace System.Runtime.Versioning
+{
+    public abstract class OSPlatformAttribute : Attribute
```

Comment:
Will these attributes exist in reference assemblies only?

Reply:
We should spec that indeed. What does Xamarin do today? Generally speaking, we don't strip most of the attributes. Leaving them in allows runtime light-up, but I can see them causing bloat.

Reply:
Xamarin doesn't have reference assemblies, there's a single implementation assembly which contains all the availability attributes. They're all removed by the managed linker at build time (if the linker is enabled).